lib/tree_construction/location_info_mixin.js

```diff
@@ -10,7 +10,19 @@ var $ = HTML.TAG_NAMES;
 
 
 function setEndLocation(element, endTagToken) {
-    if (element.__location)
+    if(!element.__startTagLocation && element.__location) {
+        element.__startTagLocation = {
+            start: element.__location.start,
+            end: element.__location.end
+        };
+    }
+    if(endTagToken.location && endTagToken.type === Tokenizer.END_TAG_TOKEN && element.tagName === endTagToken.tagName) {
```

Comment:
Cool. Maybe add brief comments with those two cases next to the clauses that check against them.